GLTFLoader: Fix node parse

takahirox opened the pull request:
Fixed #25332.

**Description**

This commit fixes a `GLTFLoader` node parse bug that can cause an infinite loop in node parse - skin parse - node parse.

The solution is adding `._loadNode()` that parses a single node without child nodes or skin to the parser. `.loadNode()` and `.loadSkin()` call `._loadNode()` to get dependent nodes. `._loadNode()` doesn't have dependency with other nodes so infinite loop can be avoided.

`._loadNode()` caches created nodes to avoid duplication because it is called from two different methods, `.loadNode()` and `.loadSkin()`.

/cc @donmccurdy

takahirox committed:
GLTFLoader: Fix node parse

This commit fixes a GLTFLoader node parse bug that
can cause an infinite loop in node parse - skin
parse - node parse.

The solution is adding ._loadNode() that parses a
single node without child nodes or skin to the parser.
.loadNode() and .loadSkin() call ._loadNode() to
get dependent nodes. ._loadNode() doesn't have
dependency with other nodes so infinite loop can
be avoided.

._loadNode() caches created nodes to avoid duplication
because it is called from two different methods,
.loadNode() and .loadSkin().

donmccurdy commented on examples/jsm/loaders/GLTFLoader.js:
```diff
@@ -3631,6 +3635,10 @@ class GLTFParser {
 
+			// Note that bones (joint nodes) may or may not be in the
+			// scene graph at this time.
+			// Child nodes are added in .loadNode() (no '_' prefix).
```
The comment might be a little misleading because this function, .loadSkin, has just been invoked by .loadNode, but not necessarily the same .loadNode call that is going to populate the joints of this skin. I would probably just remove the second sentence, "Child nodes are..." and try to clarify the method names instead. How about this?

- `.loadNode`
- `.loadNodeShallow`

With or without the underscore is fine with me!

takahirox replied:
Thanks for renaming suggestion. I know I'm bad at naming in English so renaming suggestion is always helpful to me.

I renamed `._loadNode` to `._loadNodeShallow`. I'm not sure we should really expose the method from the parser so didn't remove the underscore but we may rename later if users request.

donmccurdy replied:
Doesn't need to be a public method, agreed! :)

donmccurdy commented on examples/jsm/loaders/GLTFLoader.js:
```diff
@@ -3879,18 +3887,91 @@ class GLTFParser {
 		const nodeName = nodeDef.name ? parser.createUniqueName( nodeDef.name ) : '';
 
-		return ( function () {
+		this.nodeCache[ nodeIndex ] = ( function () {
```
if we remove the IIFE here, we could move this cache assignment about 40 lines closer to the end of the function, so it may be easier to read and see what it caches — would that be OK?

takahirox replied:
Yes, that sounds good to me.

takahirox commented on examples/jsm/loaders/GLTFLoader.js:
```diff
@@ -3879,18 +3887,91 @@ class GLTFParser {
+				skeletonPending
+			] );
+
+		}() ).then( function ( results ) {
```
This IIFE closure seems to have been introduced at some point for a certain reason but I couldn't track.

I agree with wriging without IIFE closure because it would be more readable. But do you think it's ok for you to do that in another PR? It will be easier to revert if this IIFE closure is really needed for something special reasons.

takahirox committed:
GLTFLoader: Rename _loadNode to _loadNodeShallow

This commit renames _loadNode to _loadNodeShallow.

The behavior of the method is more clarified by
the new name so remove a commend that sounds like
duplicated with the method name (and also sounds
like a bit misleading).
